### PR TITLE
Move KubeSchedulerConfiguration out of componentconfig API group 

### LINKED ISSUE #53577: kube-proxy config should move out of ComponentConfig apigroup
As per mtaufen:

... when kube-proxy moves its configuration type out of ComponentConfig, they should move the validation code along with it. Validation code should live near the type it validates, and at present the KubeProxyConfiguration type still lives in ComponentConfig.

As an example of how to remove a type from ComponentConfig (and ALL components should be removing their configuration types from this api group, by the way), see what we did for KubeletConfiguration: #49051 followed by #50789.

### PR DESCRIPTION
**Pull request by guangxuli**
**What this PR does / why we need it**:
move KubeSchedulerConfigurationout of componentconfig API group

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:  
fixes #53577

**Special notes for your reviewer**:
none

**Release note**:
```release-note
none
```

### CODE REVIEW

**Review by bsalamat — CHANGES_REQUESTED** [review]
We should wait until #52562 is merged before moving forward with this change.

**Comment by guangxuli** [comment]
@bsalamat sure, i will always watch that PR.

**Comment by guangxuli** [comment]
@k8s-merge-robot  No plan to rebase until relevant PR merge.

**Comment by xiangpengzhao** [comment]
#52562 has been merged, let's push this forward.
/hold cancel

File: plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go
Comment (ncdc) [inline plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go]: This should probably either be scheduler.config.k8s.io or kubescheduler.config.k8s.io. Given that for kube-proxy we have kubeproxy.config.k8s.io, I think kubescheduler.config.k8s.io makes sense. cc @mtaufen @mikedanese @lavalamp @liggitt @jbeda

Comment (liggitt) [reply plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go]: > I think kubescheduler.config.k8s.io makes sense

+1

**Comment by guangxuli** [comment]
good news,I will push this forward today.

**Comment by luxas** [comment]
PTAL @kubernetes/sig-scheduling-pr-reviews @kubernetes/sig-api-machinery-pr-reviews  @kubernetes/api-approvers

**Comment by guangxuli** [comment]
hmm, hi @luxas there is a confusing when removing the scheduler configuration from componentconfig. `LeaderElectionConfiguration` is used both by controller and schduler configuration(exactly `KubeSchedulerLeaderElectionConfiguration`).  when after removing the `KubeSchedulerLeaderElectionConfiguration` to `pluging/pkg/scheduler/apis/schedulerconfig/`  and change the defination of KubeSchedulerLeaderElectionConfiguration
 **from**  : 

```
 type KubeSchedulerLeaderElectionConfiguration struct {
        LeaderElectionConfiguration `json:",inline"`
	// LockObjectNamespace defines the namespace of the lock object
	LockObjectNamespace string `json:"lockObjectNamespace"`
	// LockObjectName defines the lock object name
	LockObjectName string `json:"lockObjectName"`
}
```

**to:** 
```
type KubeSchedulerLeaderElectionConfiguration struct {
	componentconfigv1alpha1.LeaderElectionConfiguration `json:",inline"`
	// LockObjectNamespace defines the namespace of the lock object
	LockObjectNamespace string `json:"lockObjectNamespace"`
	// LockObjectName defines the lock object name
	LockObjectName string `json:"lockObjectName"`
}
```
the difference is `LeaderElectionConfiguration` used by`KubeSchedulerLeaderElectionConfiguration` defined in another package(componentconfig), so this will cause the abnormal object type conversion file(**/kubernetes/plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/zz_generated.conversion.go**):

`// TODO: Inefficient conversion - can we improve it?`

```
func autoConvert_v1alpha1_KubeSchedulerLeaderElectionConfiguration_To_schedulerconfig_KubeSchedulerLeaderElectionConfiguration(in *KubeSchedulerLeaderElectionConfiguration, out *schedulerconfig.KubeSchedulerLeaderElectionConfiguration, s conversion.Scope) error {
	// TODO: Inefficient conversion - can we improve it?
	if err := s.Convert(&in.LeaderElectionConfiguration, &out.LeaderElectionConfiguration, 0); err != nil {
		return err
	}
	out.LockObjectNamespace = in.LockObjectNamespace
	out.LockObjectName = in.LockObjectName
	return nil
}
```
I think this abnormal file will be result in a verification error and bazel test error, just like : 

`W1110 08:02:46.982] F1110 08:02:46.980926   24463 server.go:342] unable to apply config defaults: converting (componentconfig.LeaderElectionConfiguration) to (v1alpha1.LeaderElectionConfiguration): type names don't match (bool, ), and no conversion 'func (bool, *bool) error' registered.`

I have spent lot of time to fix this test failure :sweat_smile: : , could you give me some suggestion? thank a lot @luxas

**Comment by guangxuli** [comment]
Also expect guys @ncdc @liggitt @bsalamat  feedback, thanks.

**Comment by guangxuli** [comment]
ping @ncdc

**Comment by ncdc** [comment]
I'm sorry, but I need to focus on some other things at the moment. I trust the other reviewers will be able to move this forward. Also cc @ironcladlou

**Comment by guangxuli** [comment]
got it, thanks @ncdc :)

**Comment by ironcladlou** [comment]
@guangxuli 

Would it make sense to move `LockObjectNamespace` and `LockObjectName` to fields of `componentconfigv1alpha1.LeaderElectionConfiguration`? Then we could delete `KubeSchedulerLeaderElectionConfiguration`. I don't think those two fields are necessarily specific to scheduler configuration and might be useful to any controller dealing with leader elections.

**Comment by guangxuli** [comment]
thanks @ironcladlou 
hmm, i am not sure whether the two fields(`LockObjectNamespace/LockObjectName`) is needed necessary .
but seems that scheduler need `LeaderElectionConfiguration`, if we remove `KubeSchedulerLeaderElectionConfiguration`, how to config `LeaderElectionConfiguration`, does it works well like before. Not sure if i understand correctly. :)

File: plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go
Comment (anfernee) [inline plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go]: s/use/used/

File: plugin/pkg/scheduler/apis/schedulerconfig/register.go
Comment (anfernee) [inline plugin/pkg/scheduler/apis/schedulerconfig/register.go]: is this TODO needed?

File: plugin/pkg/scheduler/apis/schedulerconfig/register.go
Comment (anfernee) [inline plugin/pkg/scheduler/apis/schedulerconfig/register.go]: ditto.

File: plugin/pkg/scheduler/apis/schedulerconfig/register.go
Comment (guangxuli) [reply plugin/pkg/scheduler/apis/schedulerconfig/register.go]: per https://github.com/kubernetes/kubernetes/pull/53645 there is no response from @deads2k , i will check this carefully later. thanks.

File: plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go
Comment (guangxuli) [reply plugin/pkg/scheduler/apis/schedulerconfig/v1alpha1/register.go]: will do, tks. :)

**Comment by bsalamat** [comment]
Let me know when this is ready for review.

**Comment by guangxuli** [comment]
@bsalamat hmm, there is a problem in this pr proceeding, as i have described, not sure where **LeaderElectionConfiguration** should be placed. because it is refer by both controller and scheduler.
if refer **LeaderElectionConfiguration** in scheduler by **componentconfig.LeaderElectionConfiguration** there is a bazel test error.

**Comment by guangxuli** [comment]
@bsalamat any idea?

**Comment by guangxuli** [comment]
per https://github.com/kubernetes/kubernetes/issues/54318, Seems creating a new common package to place all common configurations make sense.

**Comment by resouer** [comment]
Guangxu, I saw your bazel failure is caused by `openapi.GetOpenAPIDefinitions` is not defined, how it related to `componentconfig.LeaderElectionConfiguration` in https://github.com/kubernetes/kubernetes/pull/54211#issuecomment-356792226 may I ask?

**Comment by guangxuli** [comment]
@resouer yes, I agree, i am checking  whether the problem(openapi.GetOpenAPIDefinitions) is a new problem caused by my rebase.

**Comment by timothysc** [comment]
I'm generally a +1 to this and would like to see this move forwards in 1.10.  This will greatly help the sig-cluster-lifecycle effort going forwards.  

/cc @roberthbailey

**Comment by guangxuli** [comment]
In `KubeSchedulerLeaderElectionConfiguration` , i define LeaderElectionConfiguration by `componentconfig.LeaderElectionConfiguration`, same in internal type and external type(v1alpha1).

the auto generated code in 

```
func autoConvert_schedulerconfig_KubeSchedulerLeaderElectionConfiguration_To_v1alpha1_KubeSchedulerLeaderElectionConfiguration(in *schedulerconfig.KubeSchedulerLeaderElectionConfiguration, out *KubeSchedulerLeaderElectionConfiguration, s conversion.Scope) error {
	// TODO: Inefficient conversion - can we improve it?
	if err := s.Convert(&in.LeaderElectionConfiguration, &out.LeaderElectionConfiguration, 0); err != nil {
		return err
	}
	out.LockObjectNamespace = in.LockObjectNamespace
	out.LockObjectName = in.LockObjectName
	return nil
}
```


vefify logs:

```
> I0118 08:21:42.715]     cmd/genyaml
> W0118 08:24:23.726] F0118 08:24:23.725974   29834 server.go:341] unable to apply config defaults: converting (componentconfig.LeaderElectionConfiguration) to (v1alpha1.LeaderElectionConfiguration): type names don't match (bool, ), and no conversion 'func (bool, *bool) error' registered.
```

@resouer do you know how to handle this issue, i have confused by this for several days. :(. 
i try many ways to handle this, the verify failure still exists.

@mtaufen @timothysc  any suggestion?

**Comment by ncdc** [comment]
The internal LeaderElectionConfiguration has `LeaderElect bool`, while its `v1alpha1` counterpart has `LeaderElect *bool`. This is what is causing the problem. I believe you'll need to write a custom conversion function for this struct. Right @sttts @deads2k @liggitt?

**Comment by ncdc** [comment]
Or just change the fields to match (if that's acceptable and won't cause any backward compat issues).

**Comment by liggitt** [comment]
those should be defined:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/conversion.go#L142-L159

those are used in the autogenerated conversions for the current component config

**Comment by timothysc** [comment]
@guangxuli could you separate your PR into 2 commits. 

- code changes 
- generated updates

**Comment by guangxuli** [comment]
@timothysc sure, i just want to record the modification on fixing the verify test failure.  Will squash the debug commit  when all tests happy.

**Comment by guangxuli** [comment]
@bsalamat @luxas 
i this PR is ready to get review.

**Comment by liggitt** [comment]
> thanks so much @liggitt @ncdc that's very very helpful.

sorry, my point was that the bool pointer conversions should already be usable... you shouldn't have to redefine them. componentconfig used them and autogeneration picked them up without redefining.

**Comment by guangxuli** [comment]
@liggitt oh, yes, you are right, will modify that.

**Comment by php-coder** [comment]
@guangxuli ping

**Comment by guangxuli** [comment]
@php-coder will rebase this PR, i just finished the Spring Festival to work today.

**Comment by php-coder** [comment]
@guangxuli Needs rebase.

**Comment by ncdc** [comment]
No matter what, it needs to move out of `pkg/apis/componentconfig`. The question that @sttts is raising is if it should move under `cmd/kube-scheduler` or `pkg/scheduler`.

Do we have well defined guidelines for putting code in `cmd` vs `pkg`?

**Comment by guangxuli** [comment]
yes, there is seems no final decision for where we put the scheduler configurations  **cmd/scheduler** or **pkg/scheduler**.

**Comment by liggitt** [comment]
prefer pkg. c.f.:

k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig
k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig

**Comment by liggitt** [comment]
I see @sttts argued for moving config types into cmd packages in https://github.com/kubernetes/kubernetes/pull/58565#issuecomment-369145765

I'd expect other components to make use of the serialized config structs eventually, and would not expect them to depend on the cmd packages to do that

**Comment by ncdc** [comment]
@liggitt I think @sttts was suggesting we move the proxy and kubelet apis under cmd

**Comment by wanghaoran1988** [comment]
@guangxuli Are you still working on this?

**Comment by luxas** [comment]
FWIW, I prefer to have the componentconfig under `pkg/` for the reasons @liggitt mentioned.
It'd be great with a rebase here and a push to get this into v1.11 (even though it would still be marked alpha in v1.11)

**Comment by luxas** [comment]
Closing this one in favor of https://github.com/kubernetes/kubernetes/pull/66916 which implements the formal KEP proposal https://github.com/kubernetes/community/pull/2515
Thanks for the work here @guangxuli that paved the way for the work that is now in #66916 :smile:

**Comment by guangxuli** [comment]
@luxas :+1:

**Comment by bgrant0607** [comment]
On the issue of where to put config types, please start a thread on the kubernetes-sig-architecture mailing list.

**Pull request closed without merging**
